// pinia-plugin-persistedstate/dist/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  createNuxtPersistedState: () => createNuxtPersistedState,
  createPersistedState: () => createPersistedState,
  default: () => src_default
});

// src/pick.ts
function get(state, path) {
  return path.reduce((obj, p) => {
    return obj == null ? void 0 : obj[p];
  }, state);
}
function set(state, path, val) {
  return path.slice(0, -1).reduce((obj, p) => {
    if (!/^(__proto__)$/.test(p))
      return obj[p] = obj[p] || {};
    else
      return {};
  }, state)[path[path.length - 1]] = val, state;
}
function pick(baseState, paths) {
  return paths.reduce((substate, path) => {
    const pathArray = path.split(".");
    return set(substate, pathArray, get(baseState, pathArray));
  }, {});
}

// src/index.ts
function createPersistedState(factoryOptions = {}) {
  return function(context) {
    var _a, _b, _c, _d;
    const {
      options: { persist },
      store
    } = context;
    if (!persist)
      return;
    const {
      storage = (_a = factoryOptions.storage) != null ? _a : localStorage,
      beforeRestore = (_b = factoryOptions.beforeRestore) != null ? _b : null,
      afterRestore = (_c = factoryOptions.afterRestore) != null ? _c : null,
      serializer = (_d = factoryOptions.serializer) != null ? _d : {
        serialize: JSON.stringify,
        deserialize: JSON.parse
      },
      key = store.$id,
      paths = null
    } = typeof persist != "boolean" ? persist : {};
    beforeRestore == null ? void 0 : beforeRestore(context);
    try {
      const fromStorage = storage.getItem(key);
      if (fromStorage)
        store.$patch(serializer.deserialize(fromStorage));
    } catch (_error) {
    }
    afterRestore == null ? void 0 : afterRestore(context);
    store.$subscribe((_mutation, state) => {
      try {
        const toStore = Array.isArray(paths) ? pick(state, paths) : state;
        storage.setItem(key, serializer.serialize(toStore));
      } catch (_error) {
      }
    }, { detached: true });
  };
}
function createNuxtPersistedState(useCookie, factoryOptions) {
  return createPersistedState(__spreadValues({
    storage: {
      getItem: (key) => {
        return useCookie(key, {
          encode: (x) => x,
          decode: (x) => x
        }).value;
      },
      setItem: (key, value) => {
        useCookie(key, {
          encode: (x) => x,
          decode: (x) => x
        }).value = value;
      }
    }
  }, factoryOptions));
}
var src_default = createPersistedState();
module.exports = __toCommonJS(src_exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createNuxtPersistedState,
  createPersistedState
});
